rfcs/0039-snark-keys-management.md

```diff
@@ -0,0 +1,208 @@
+  keys for a particular binaries when deploying.
+* The caching behavior for keys is implicit and entangled with the pickles
+  proof system interface, which makes it hard for the team to inspect or adjust
+  the behavior.
+
+The goal of this RFC is to resolve these issues. At a high-level, the proposed
+steps to do this are:
+* Make keys identifiable from the file contents.
+* Add a tool for generating keys outside of the build process.
+* Allow new keys to be specified explicitly.
+* Make key generation explicit, removing it from the build process.
+
+## Detailed design
+[detailed-design]: #detailed-design
+
+### Make keys identifiable from the file contents
+
+In order to confirm that loaded keys are compatible with a given configuration,
+we need to have some representation of the configuration embedded with the keys
+files. We propose that every key (and URS) file contains a header in minified
+JSON format before the actual contents, specifying
+* the constraint constants, as defined in the runtime configuration
+* the SHA commit identifiers for the current commit in the coda, marlin and
+  zexe repos used when generating the keys
+* the length of the binary data following the header, so that we can validate
+  that the file was successfully downloaded in full, where applicable
+* the date when the file's contents were created
+* the type of the file's contents
+* the constraint system hash (for keys) or domain size (for URS)
+* any other input information
+  - e.g. constraint system hash for the transaction snark wrapped by a
+    particular blockchain snark
+* a version number, so that we can add or modify fields
+* the identifying hash, which should match the hash part of the filename for
+  generated files
+
+For example, a header for the transaction snark key file might look like:
+```json
+{ "header_version": 1
+, "kind":
+    { "type": "step_proving_key"
+    , "identifier": "transaction_snark_base" }
+, "constraint_constants":
+    { "c": 8
+    , "ledger_depth": 14
+    , "work_delay": 2
+    , "block_window_duration_ms": 180000
+    , "transaction_capacity": {"txns_per_second_x10": "2"}
+    , "coinbase_amount": "200"
+    , "supercharged_coinbase_factor": 2
+    , "account_creation_fee": "0.001" }
+, "commits":
+    { "coda": "COMMIT_SHA_HERE"
+    , "marlin": "COMMIT_SHA_HERE"
+    , "zexe": "COMMIT_SHA_HERE" }
+, "length": 1000000000
+, "date": "1970-01-01 00:00:00"
+, "constraint_system_hash": "MD5_HERE"
+, "identifying_hash": "HASH_HERE" }
+```
+
+This will allow us to identify files from S3 or in a cache directory by
+examining only the header, making it easier to remove outdated keys and to
+find or enumerate keys.
+
+The majority of the work here is adjusting the rust code so that the header can
+be generated from OCaml and written to the same file as its contents.
+
+### Add a tool for generating keys outside of the build process
+
+This tool could be an `internal`/`advanced` subcommand of the main executable
```

Comment:
There is a lot of value in having this as a distinct executable, especially in making sure that most tests/jobs are not dependent on building coda.exe. Is the time/effort cost of splitting it out much higher? Alternatively, do we save effort by adding this command / these features to runtime_genesis_ledger.exe instead? Or src/lib/snark_keys/gen_keys/gen_keys.exe which is already focused on snark key generation/management?

Reply:
We can build a separate executable that sits at the same layer of the dependency graph as snark_keys does (and library-ise the command so that it can still be exposed from the CLI too), if that would be helpful. I will update the text to add this, and it should be no extra work to do it this way.

I'll note though that the snark keys depend on code quite far up the coda.exe dependency chain, so we'll have to recompile it for each commit, and we likely won't end up saving much time vs using compiling and using coda.exe directly.